File: core/AllElementsInTwoBinarySearchTrees/AllElementsInTwoBinarySearchTrees.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;

public class AllElementsInTwoBinarySearchTrees {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private void dfs(TreeNode node, List<Integer> nums) {
        if (node != null) {
            dfs(node.left, nums);
            nums.add(node.val);
            dfs(node.right, nums);
        }
    }

    public List<Integer> getAllElements(TreeNode root1, TreeNode root2) {
        List<Integer> nums1 = new ArrayList<>(), nums2 = new ArrayList<>();
        dfs(root1, nums1);
        dfs(root2, nums2);

        List<Integer> res = new ArrayList<>();
        int left = 0, right = 0;
        while (left < nums1.size() || right < nums2.size()) {
            int num1 = left < nums1.size() ? nums1.get(left) : 1000000;
            int num2 = right < nums2.size() ? nums2.get(right) : 1000000;
            if (num1 <= num2) {
                res.add(num1);
                left++;
            } else {
                res.add(num2);
                right++;
            }
        }
        return res;
    }
}

File: core/AllElementsInTwoBinarySearchTrees/AllElementsInTwoBinarySearchTrees.py
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def getAllElements(root1: TreeNode, root2: TreeNode) -> List[int]:
    def dfs(node, nums):
        if not node:
            return
        dfs(node.left, nums)
        nums.append(node.val)
        dfs(node.right, nums)

    nums1, nums2, res = [], [], []
    dfs(root1, nums1)
    dfs(root2, nums2)
    len1, len2, left, right = len(nums1), len(nums2), 0, 0
    while left < len1 or right < len2:
        num1 = nums1[left] if left < len1 else 1000000
        num2 = nums2[right] if right < len2 else 1000000
        if num1 <= num2:
            res.append(num1)
            left += 1
        else:
            res.append(num2)
            right += 1

    return res



